compiler/optimizer/VPConstraint.cpp

```diff
@@ -610,11 +610,21 @@ TR_YesNoMaybe TR::VPClassType::isJavaLangClassObject()
    // it would make this look just like Class.class.
    if (_len == 17 && strncmp(_sig, "Ljava/lang/Class;", 17) == 0)
       return TR_maybe;
+
+   // If the type is java/lang/Object or an interface that is implemented
+   // by java/lang/Class, this may be the java/lang/Class object itself.
+   //
+   // java/lang/constant/Constable, java/lang/invoke/TypeDescriptor
+   // and java/lang/invoke/TypeDescriptor$OfField were introduced in JDK 12
+   //
    if ((_len == 18 && strncmp(_sig, "Ljava/lang/Object;", 18) == 0) ||
          (_len == 22 && strncmp(_sig, "Ljava/io/Serializable;", 22) == 0) ||
          (_len == 36 && strncmp(_sig, "Ljava/lang/reflect/AnnotatedElement;", 36) == 0) ||
          (_len == 38 && strncmp(_sig, "Ljava/lang/reflect/GenericDeclaration;", 38) == 0) ||
-         (_len == 24 && strncmp(_sig, "Ljava/lang/reflect/Type;", 24) == 0))
+         (_len == 24 && strncmp(_sig, "Ljava/lang/reflect/Type;", 24) == 0) ||
+         (_len == 30 && strncmp(_sig, "Ljava/lang/constant/Constable;", 30) == 0) ||
+         (_len == 33 && strncmp(_sig, "Ljava/lang/invoke/TypeDescriptor;", 33) == 0) ||
+         (_len == 41 && strncmp(_sig, "Ljava/lang/invoke/TypeDescriptor$OfField;", 41) == 0))
       return TR_maybe;
    return TR_no; // java.lang.Class is final and is the direct subclass of Object.
```

Comment:
Could we create an assert that guards against this in the future ? e.g. use the front end apis to get the opaque class blocks for all the interfaces implemented by `java/lang/Class` and compare vs the "known set of interfaces" ? 

Maybe this could even be used to create an alternate fix that does not depend on the signature matching of a fixed set, but instead a signature matching based on the set of opaque class blocks for all the interfaces implemented by `java/lang/Class` as per a front end query that we create. 

If this has merit, I would be okay with you rolling it into this PR or alternately, choosing to do it via a separate PR. Please let me know your thoughts so that I can decide whether to run testing or not.

Reply:
That is fine with me. I'll start testing, thanks.